text/3351-relax-const-restrictions.md

```diff
@@ -0,0 +1,99 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC allows `const fn` to exhibit different behavior during constant evaluation and runtime. While such a difference is often undesirable, it is not considered to be undefined behavior.
+
+If a `const fn` is able to detect whether it has been called during constant evaluation or at runtime (either through an intrinsic or a future language feature), then it is allowed to exhibit different behavior. Also, a `const fn` called at runtime can do anything a normal function can do, with no additional restrictions applied to it. It could open a file, call a system randomness API or gracefully exit the program. 
+
+At the time of writing, there is no way for a function to detect whether it was called at runtime or during constant evaluation in stable Rust and this RFC is not concerned with adding any, but it unblocks future RFCs for adding this capability.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Each execution of a function stands in a particular "constness context". A `const fn` is executed in a const context if it was called inside another `const fn` that was executed in a const context or if it was called in one of the following places:
+
+- `const` initializers (`const X: _ = CONST;`)
+- `static` initializers (`static X: _ = CONST;`)
+- array lengths (`[T; CONST]`)
+- enum discriminants (`enum A { B = CONST }`)
+- inline-const block (`const { CONST }`)
+- const generic arguments (`function::<CONST>()`)
+
+This list may be extended by future language features.
+
+All other calls to a `const fn` (for example in `main`) are in a runtime context.
+
+We can therefore say that statically, code can either be:
+- Always in a const context (code inside one of the places listed above)
+- Maybe in a const context (`const fn`), where the context can differ between calls depending on the call-site
+- Always in a runtime const context (non-`const` functions)
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a const or runtime context. Language features and standard library functions may also differ in behavior depending on the context they have been used or called in, though the Rust language and standard library will explicitly document such behavioral differences.
+
+This makes the context of a function observable behavior.
+
+If a `const fn` is called in a runtime context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+A `const fn` being called in a const context will still be required to be deterministic, as this is required for type system soundness. This invariant is required by the compiler and cannot be broken, even with unsafe code.
```

Comment:
This means that, for example, CTFE can't use host floats for anything that might be NAN, because LLVM might change the NANs when run at different times, right?

Reply:
Yes.